File: java/0590.N叉树的后序遍历.java
class Solution {
    // 1. O(n) t:3ms(37%) O(n) m:42.5M(5%) 迭代
    public List<Integer> postorder(Node root) {
        Deque<Node> stack = new ArrayDeque<>();
        List<Integer> ans = new ArrayList<>();
        if (root == null) return ans;
        stack.addLast(root);
        while (!stack.isEmpty()) {
            Node node = stack.pollLast();
            if (node == null) continue;
            for (Node n : node.children) {
                stack.addLast(n);
            }
            ans.add(node.val);
        }
        Collections.reverse(ans);
        return ans;
    }
}

File: python/0590.N叉树的后序遍历.py
class Solution:
    # 1. O(n) t:52ms(63%) O(n) m:16.8M(55%) 迭代
    def postorder(self, root: 'Node') -> List[int]:
        if root is None: return []
        stack, ans = root.children, [root.val]
        while stack:
            node = stack.pop()
            if node.children:
                stack.extend(node.children)
            ans.append(node.val)
        return ans[::-1]
    
    # 2. O(n) t:56ms(39%) m:16.9M(40%) 递归
    def postorder(self, root: 'Node') -> List[int]:
        if root is None: return []
        ans = []
        def dfs(node):
            nonlocal ans
            for n in node.children:
                dfs(n)
            ans.append(node.val)
        
        dfs(root)
        return ans